test/e2e/hot-reload.test.js

```diff
@@ -96,75 +94,65 @@ describe('hot-reload', () => {
-          .end((err, res) => {
-            assert(err);
-            assert(res.clientError);
-            assert.equal(res.statusCode, 404);
-            done();
-          });
-      });
+    beforeEach(function (done) {
+      watcher = chokidar.watch(testGatewayConfigPath, { awaitWriteFinish: true });
+      watcher.on('ready', done);
+    });
 
-      watcher.on('ready', () => {
-        // remove key-auth policy
-        testGatewayConfigData.pipelines.adminAPI.policies.shift();
+    afterEach(function () {
+      watcher.close();
+    });
 
-        fs.writeFile(testGatewayConfigPath, yaml.dump(testGatewayConfigData), (err) => {
-          if (err) {
-            throw err;
-          }
+    describe('reloads valid gateway.config.yml', function () {
+      it('will respond with a 404 - proxy policy', function (done) {
+        this.timeout(TEST_TIMEOUT);
+        watcher.once('change', (evt) => {
+          setTimeout(() => {
+            request
+              .get(`http://localhost:${originalGatewayPort}`)
+              .end((err, res) => {
+                assert(err);
+                assert(res.clientError);
+                assert.equal(res.statusCode, 404);
+                done();
+              });
+          }, GATEWAY_STARTUP_WAIT_TIME);
         });
-      });
-    }).timeout(10000);
-
-    it('uses previous config on reload of invalid gateway.config.yml', done => {
-      const watchOptions = {
-        awaitWriteFinish: true
-      };
-
-      const watcher = chokidar.watch(testGatewayConfigPath, watchOptions);
-      watcher.once('change', (evt) => {
-        request
-          .get(`http://localhost:${originalGatewayPort}`)
-          .end((err, res) => {
-            assert(err);
-            assert(res.clientError);
-            assert.equal(res.statusCode, 404);
-            done();
-          });
-      });
 
-      watcher.on('ready', () => {
-        // make config invalid
-        delete testGatewayConfigData.pipelines;
+        testGatewayConfigData.pipelines.adminAPI.policies.shift();
+        fs.writeFileSync(testGatewayConfigPath, yaml.dump(testGatewayConfigData));
+      });
+    });
 
-        fs.writeFile(testGatewayConfigPath, yaml.dump(testGatewayConfigData), (err) => {
-          if (err) {
-            throw err;
-          }
+    describe('uses previous config on reload of invalid gateway.config.yml', function () {
+      it('will respond with 404 - empty proxy', function (done) {
+        this.timeout(TEST_TIMEOUT);
+        watcher.once('change', () => {
+          request
+            .get(`http://localhost:${originalGatewayPort}`)
+            .end((err, res) => {
+              assert(err);
+              assert(res.clientError);
+              assert.equal(res.statusCode, 404);
+              done();
+            });
         });
+        // make config invalid
+        fs.writeFileSync(testGatewayConfigPath, yaml.dump("ahahahaha-0921i3"));
```

Comment:
This is a bit tricky - it is important to make sure the event handler is in place before writing on the file. Making both async is impossible - it would be totally untestable.

Reply:
`ahahahaha-0921i3`  is valid yaml 
write pure text - `{er:t4`